src/client/datascience/notebook/notebookControllerManager.ts

```diff
@@ -0,0 +1,274 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+'use strict';
+import { inject, injectable } from 'inversify';
+import { CancellationTokenSource, EventEmitter, NotebookController, NotebookDocument, NotebookSelector } from 'vscode';
+import { IExtensionSingleActivationService } from '../../activation/types';
+import { ICommandManager, IVSCodeNotebook } from '../../common/application/types';
+import { PYTHON_LANGUAGE } from '../../common/constants';
+import { traceInfo, traceInfoIf } from '../../common/logger';
+import { IConfigurationService, IDisposableRegistry, IExtensionContext, IExtensions } from '../../common/types';
+import { noop } from '../../common/utils/misc';
+import { sendNotebookOrKernelLanguageTelemetry } from '../common';
+import { Telemetry } from '../constants';
+import { areKernelConnectionsEqual, isLocalLaunch } from '../jupyter/kernels/helpers';
+import { IKernelProvider, KernelConnectionMetadata } from '../jupyter/kernels/types';
+import { ILocalKernelFinder, IRemoteKernelFinder } from '../kernel-launcher/types';
+import { INotebookStorageProvider } from '../notebookStorage/notebookStorageProvider';
+import { PreferredRemoteKernelIdProvider } from '../notebookStorage/preferredRemoteKernelIdProvider';
+import { sendKernelTelemetryEvent, trackKernelResourceInformation } from '../telemetry/telemetry';
+import { INotebookProvider } from '../types';
+import { JupyterNotebookView } from './constants';
+import { getNotebookMetadata, isJupyterKernel, isJupyterNotebook, trackKernelInNotebookMetadata } from './helpers/helpers';
+import { VSCodeNotebookController } from './notebookExecutionHandler';
+import { INotebookControllerManager } from './types';
+/**
+ * This class tracks notebook documents that are open and the provides NotebookControllers for
+ * each of them
+ */
+@injectable()
+export class NotebookControllerManager implements INotebookControllerManager, IExtensionSingleActivationService {
```

Comment:
Might want to use `IExtensionSyncActivationService` as this has an actiavte method thats synchronous. This way things complete immediately instead of `async`.
The drawback of using `async` is because is slows down the rest, etc.